tests/test_ranges.py

```diff
@@ -964,6 +965,34 @@ def test_union_of_overlapping_ranges(self):
                 end=datetime.datetime(2000, 1, 4),
             )
 
+    class TestLocalize:
+        def test_converts_timezone(self):
```

Comment:
Can you please add a test here over DST boundaries? Specifically what happens when you have an hour range (in utc) and the start localises to a different tz-offset, and a separate test when end localises to a different tz-offset.

I'm not confident `astimezone` will do the right thing - whether that correctly adjusts the offset or not. We dealt with a similar issue of DateTimeRangeField (that would automatically localise) and it would result in empty ranges, because the post-localised value included a `fold=1` attribute indicating which *side* of the boundary it should be on, but the start and end values themselves had the same *time*.

Reply:
Very interesting stuff!

> Localising a datetime, natively, does not work well with ranges

In what way?

Some periods of time just "don't exist" in other timezones - the errors reflect that.

> that should be 0 seconds - but the timedelta module doesn't consider offsets important

I'm not sure this example is realistic: 2AM doesn't actually exist at that point in time.

Arguably [`datetime` itself should error on creation if the point in time doesn't exist](https://peps.python.org/pep-0495/#the-fold-attribute):

> ...instances that fall in the spring-forward gap. Such instances don’t represent any valid time

But this isn't just [Garbage In, Garbage Out](https://en.wikipedia.org/wiki/Garbage_in,_garbage_out), as `timedelta` does emit invalid datetimes:

```python
datetime(2024, 10, 6, 2, tzinfo=SYD) + timedelta(hours=1)
=> datetime.datetime(2024, 10, 6, 2, 0, tzinfo=...)  # <== WRONG
```

Overall this is a problem with `datetime` itself. If it weren't a problem, I don't think we
would be having this discussion, or any qualms about this function.

🐮 For now I've done a PR to document this pitfall: https://github.com/octoenergy/xocto/pull/180

🐮 Maybe we should wrap `datetime` so it actually behaves correctly 🐉 💥.

Reply:
(I've fixed up the new tests into the second commit, which adds the function under test.)

Reply:
One more thing - the other side of DST where time goes backwards is similarly bad, and isn't an example of garbage in.

```
In [36]: localtime.as_localtime(datetime(2024, 4, 6, 17, tzinfo=UTC), tz=SYD) - localtime.as_localtime(datetime(2024, 4, 6, 16, tzinfo=UTC), tz=SYD)
Out[36]: datetime.timedelta(seconds=3600)

In [37]: localtime.as_localtime(datetime(2024, 4, 6, 17, tzinfo=UTC), tz=SYD) - localtime.as_localtime(datetime(2024, 4, 6, 15, tzinfo=UTC), tz=SYD)
Out[37]: datetime.timedelta(seconds=3600)
```

Here, we see two 2 hours of UTC being localised to 1 hour of duration, because resolving it to "SYD" is the 2am *after* the wind-back, and hence the fold attribute is set:

```
In [39]: localtime.as_localtime(datetime(2024, 4, 6, 16, tzinfo=UTC), tz=SYD)
Out[39]: datetime.datetime(2024, 4, 7, 2, 0, fold=1, tzinfo=zoneinfo.ZoneInfo(key='Australia/Sydney'))
```

Timezone *names* aren't helpful for comparing durations over DST boundaries. It's all.. annoying and hard.

Reply:
> Should dict[Any, datetime] support localize(tz=...)?

Well, no, but I'd say that's more because `dict` is generic, where `FiniteDatetimeRange` is specific.

> the other side of DST where time goes backwards is similarly bad

Hmm...thought provoking. To make the localization explicit:

- 2024-04-06 15:00 UTC => 2024-04-07 02:00 SYD
- 2024-04-06 16:00 UTC => 2024-04-07 02:00 SYD
- 2024-04-06 17:00 UTC => 2024-04-07 03:00 SYD

> Here, we see two 2 hours of UTC being localised to 1 hour of duration

Is it wrong though? SYD is losing an hour after all, so that has to have an effect somewhere.